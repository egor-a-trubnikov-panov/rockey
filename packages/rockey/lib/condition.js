import interpolateString from './utils/interpolateString';

const condition = conditionFunc => (...args) => {
  if (conditionFunc()) {
    return interpolateString(...args);
  } else {
    return null;
  }
};

export default condition;
