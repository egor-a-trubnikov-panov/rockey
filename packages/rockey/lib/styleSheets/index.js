import StyleSheet from './StyleSheet';

const sheet = new StyleSheet();

let insertWithQueue = false;

export const insert = precss => {
  if (insertWithQueue) {
    sheet.insertWithQueue(precss);
  } else {
    sheet.insert(precss);
  }
};

export const getRules = css => sheet.getRules();
export const clearStyles = css => sheet.clear();

export const speedy = () => sheet.enableSpeedy();
export const useQueue = () => {
  insertWithQueue = true;
};

export default insert;
