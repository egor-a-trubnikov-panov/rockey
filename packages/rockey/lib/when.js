import isFunction from 'lodash/isFunction';
import interpolateString from './utils/interpolateString';

const when = (...args) => {
  let displayName = null;
  let mixinFunc = null;

  if (args.length === 1) {
    mixinFunc = args[0];
  } else {
    displayName = args[0];
    mixinFunc = args[1];
  }

  return (strings, ...values) => {
    let inline = null;
    let inlineAsFunction = null;

    if (isFunction(strings)) {
      inlineAsFunction = strings;
    } else {
      inline = interpolateString(strings, ...values);
    }

    const anonWhen = (...props) => {
      if (mixinFunc(...props)) {
        return inlineAsFunction ? inlineAsFunction(...props) : inline;
      }

      return null;
    };

    if (displayName) {
      anonWhen.displayName = displayName;
    } else {
      anonWhen.displayName = `anonWhen`;
    }

    return anonWhen;
  };
};

export default when;
