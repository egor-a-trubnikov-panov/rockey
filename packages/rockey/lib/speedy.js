import { speedy } from './styleSheets';
speedy();
