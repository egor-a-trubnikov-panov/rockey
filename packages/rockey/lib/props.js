import when from './when';

const props = name => {
  return when(name, () => true);
};

export default props;
