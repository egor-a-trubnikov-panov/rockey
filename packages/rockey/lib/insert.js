import { staticRule } from './rule';

export const insert = (strings, ...values) => {
  const css = staticRule(strings, ...values);
  css.getClassList();
};

export default insert;
