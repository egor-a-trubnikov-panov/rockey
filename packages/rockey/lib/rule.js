import createParser from 'rockey-css-parse';
import isString from 'lodash/isString';
import * as classnames from 'rockey-css-parse/utils/classnames';

import {
  getClassName,
  getAnimationName,
  getMixinClassName,
} from './css/getClassName';

import insert from './styleSheets';
import wrapTemplateStrings from './utils/wrapTemplateStrings';

import vendorPrefix from './plugins/vendorPrefix';
import validateCSSRule from './plugins/validateCSSRule';

const plugins = [];

plugins.push(vendorPrefix());

// if (process.env.NODE_ENV !== 'production') {
//   plugins.push(validateCSSRule());
// }

const parse = createParser({
  getClassName,
  getMixinClassName,
  getAnimationName,
  plugins,
});

const getSelector = classnames.getSelector(getClassName);

export default function rule(raw, ...values) {
  // export default function rule(...args) {
  let parent = null;
  let cached = null;
  let inserted = false;
  let isAlreadyParsed = false;
  let addedMixins = [];

  if (Array.isArray(raw)) {
    isAlreadyParsed = !isString(raw[0]);
  } else if (isString(raw)) {
  } else {
    throw new Error('rockey-rule: wrong call type');
  }

  const key = JSON.stringify({
    raw,
    values,
  });

  return {
    wrapWith(displayName) {
      if (cached) {
        throw new Error(
          'Can not wrap because rule is already parsed and inserted'
        );
      }

      if (isAlreadyParsed) {
        const selector = getSelector(displayName);

        raw.forEach(precss => {
          precss.selector = precss.selector.map(p => `${selector} ${p}`);
          precss.root = [displayName];

          precss.mixins.forEach(m => {
            m.updateSelector(ms => `${selector} ${ms}`);
          });
        });
      } else {
        raw = wrapTemplateStrings(displayName, raw);
      }
    },

    addParent(rule) {
      parent = rule;
    },

    parse() {
      let precss = null;
      let classList = null;

      if (!cached) {
        if (isAlreadyParsed) {
          precss = raw;
          classList = {};

          raw.forEach(part => {
            part.root.forEach(c => {
              if (!classList[c]) {
                classList[c] = getClassName(c);
              }
            });
          });
        } else {
          let parsed = parse(raw, ...values);

          precss = parsed.precss;
          classList = parsed.classList;
        }

        if (addedMixins.length) {
          Object.keys(classList).forEach(displayName => {
            addedMixins.forEach(addedMixin => {
              const css = rule`
                ${displayName} {
                  ${addedMixin}
                }
              `;

              css.parse().precss.forEach(p => {
                precss.push(p);
              });

              // precss = precss.concat(css.parse().precss);
            });
          });
        }

        cached = {
          precss,
          classList,
        };
      } else {
        precss = cached.precss;
        classList = cached.classList;
      }

      return { precss, classList };
    },

    transform(transformFunc) {
      let precss = {};

      if (isAlreadyParsed) {
        precss = raw;
      } else {
        const parsed = this.parse();
        precss = parsed.precss;
      }

      const tree = {};

      precss.forEach(part => {
        part.root.forEach(key => {
          if (!tree[key]) {
            tree[key] = [];
          }
          tree[key].push(part);
        });
      });

      return transformFunc(tree, transformed => {
        return rule(Array.isArray(transformed) ? transformed : [transformed]);
      });
    },

    addMixins(mixins) {
      addedMixins = addedMixins.concat(mixins);
    },

    getClassList(props = {}) {
      let parentClasses = [];

      if (parent) {
        const parentClassList = parent.getClassList(props);
        Object.keys(parentClassList).forEach(name => {
          parentClasses.push(parentClassList[name]);
        });
      }

      const { classList, precss } = this.parse();

      if (!inserted) {
        if (precss.length) {
          insert(precss);
        }

        inserted = true;
      }

      let resultClassList = {};

      if (parent) {
        let parentClassList = parentClasses.join(' ');
        if (parentClassList) {
          Object.keys(classList).forEach(name => {
            resultClassList[name] = `${classList[name]} ${parentClassList}`;
          });
        }
      } else {
        resultClassList = Object.assign({}, classList);
      }

      let mixinsPrecss = [];

      precss.forEach(pre => {
        if (pre.mixins) {
          pre.mixins.forEach(mixin => {
            const result = mixin(props);

            if (result) {
              if (result.className) {
                pre.root.forEach(c => {
                  resultClassList[c] += ` ${result.className}`;
                });
              }

              if (result.precss) {
                result.precss.forEach(p => {
                  mixinsPrecss.push(p);
                });
              }
            }
          });
        }
      });

      if (mixinsPrecss.length) {
        insert(mixinsPrecss);
      }

      return resultClassList;
    },
  };
}
