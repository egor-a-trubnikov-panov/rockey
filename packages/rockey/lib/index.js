import condition from './condition';
import rule from './rule';
import props from './props';
import insert from './insert';
import when from './when';

const rockey = {
  condition,
  rule,
  insert,
  when,
};

export default rockey;
export { condition };
export { insert };
export { when };
export { rule };
export { props };
