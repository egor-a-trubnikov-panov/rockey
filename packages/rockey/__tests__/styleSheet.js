import rule, { clearStylesCache } from '../lib/rule';
import when from '../lib/when';
import { getRules, getMixins, clearStyles } from '../lib/styleSheets';

jest.mock('../lib/utils/hash', () => {
  return () => '{{ hash }}';
});

console.warn = jest.fn();

const primary = when(props => props.isPrimary)`
  color: blue
`;

test('inserted styles', () => {
  clearStyles();
  clearStylesCache();

  const css = rule`
    Button {

      font-size: 14px;
      line-height: 36px;
      box-sizing: border-box;
      text-transform: uppercase;
      border-radius: 3px;
      user-select: none;
      border: 0;
      font-weight: 500;
      letter-spacing: .010em;
      outline: 0;
      cursor: pointer;
      overflow: hidden;
      display: inline-block;
      white-space: nowrap;
      padding: 0 6px;
      margin: 6px 8px;
      min-height: 36px;
      min-width: 88px;
      text-align: center;
      transition:
        box-shadow .4s cubic-bezier(.25,.8,.25,1),
        background-color .4s cubic-bezier(.25,.8,.25,1);

      background: transparent;

      :not(:disabled):hover {
        background-color: transparent
      }

      :focus {
        outline: 0;
      }

      :disabled {
        color: #ccc;
        background: transparent;
      }

      ${primary}
    }
  `;

  expect(css.getClassList()).toEqual({
    Button: ['Button-{{ hash }}'],
  });

  expect(getRules()).toMatchSnapshot();
  expect(getMixins()).toMatchSnapshot();
});

// test('inserted styles batching', () => {
//   clearStyles();
//   clearStylesCache();
//
//   return new Promise(
//     (resolve, reject) => {
//       setTimeout(() => {
//         const css1 = rule` Button1 { color: red; }`;
//         const css2 = rule` Button2 { color: red; }`;
//         const css3 = rule` Button3 { color: red; }`;
//         // const css4 = rule` Button4 { color: red; }`;
//         // const css5 = rule` Button5 { color: red; }`;
//
//         css1.getClassList();
//         expect(getRules().length).toEqual(1);
//
//         css2.getClassList();
//         css3.getClassList();
//         expect(getRules().length).toEqual(1);
//
//         setTimeout(
//           () => {
//             try {
//               expect(getRules().length).toEqual(3);
//
//               resolve();
//             } catch (e) {
//               reject(e);
//             }
//           },
//           5
//         );
//       });
//     },
//     5
//   );
// });

// test('inserted styles with mixin', () => {
//   clearStyles();
//   clearStylesCache();
//
//   const css = rule`
//     Button {
//       color: purple;
//
//       ${primary}
//     }
//   `;
//
//   expect(
//     css.getClassList({
//       isPrimary: true,
//     })
//   ).toEqual({
//     Button: ['Button-{{ hash }}', 'Mixin-anonWhen-{{ hash }}-1'],
//   });
//
//
//   expect(getRules()).toMatchSnapshot();
//   expect(getMixins()).toMatchSnapshot();
//
//   return new Promise((resolve, reject) => {
//     setTimeout(() => {
//       try {
//         expect(getRules()).toMatchSnapshot();
//         expect(getMixins()).toMatchSnapshot();
//
//         resolve();
//       } catch (e) {
//         reject(e);
//       }
//     });
//   });
// });

// test('inserted above threshold', () => {
//   clearStyles();
//   clearStylesCache();
//
//   const styles = [];
//
//   for (let i = 0; i <= 200; i++) {
//     styles.push(`
//       Button${i} {
//         color: purple;
//       }
//     `
//     );
//   }
//
//   const css = rule`${styles.join('')}`;
//
//   css.getClassList();
//
//   expect(getRules().length).toBeGreaterThan(0);
//   expect(getMixins().length).toEqual(0);
//
//   expect(getRules()).toMatchSnapshot();
//
//   expect(getMixins()).toMatchSnapshot();
// });
