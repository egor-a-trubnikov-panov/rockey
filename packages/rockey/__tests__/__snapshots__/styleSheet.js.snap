// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inserted styles 1`] = `
Array [
  "
    .Button-{{ hash }}:not(:disabled):hover {
        background-color:transparent;
  }
  
    .Button-{{ hash }}:focus {
        outline:0;
  }
  
    .Button-{{ hash }}:disabled {
        color:#ccc;background:transparent;
  }
  
    .Button-{{ hash }} {
        font-size:14px;line-height:36px;box-sizing:border-box;text-transform:uppercase;border-radius:3px;user-select:none;border:0;font-weight:500;letter-spacing:.010em;outline:0;cursor:pointer;overflow:hidden;display:inline-block;white-space:nowrap;padding:0 6px;margin:6px 8px;min-height:36px;min-width:88px;text-align:center;transition:box-shadow .4s cubic-bezier(.25,.8,.25,1), background-color .4s cubic-bezier(.25,.8,.25,1);background:transparent;
  }
  ",
]
`;

exports[`inserted styles 2`] = `Array []`;
